scripts/http-jsonp-detection.nse

```diff
@@ -0,0 +1,174 @@
+used to bypass Same-origin Policy restrictions in web browsers.
+
+The script searches for callback functions in the response to detect JSONP
+endpoints. It also tries to determine callback function through URL(callback 
+function may be fully or partially controllable from URL) and also tries to 
+bruteforce the most common callback variables through the URL.
+
+References : https://securitycafe.ro/2017/01/18/practical-jsonp-injection/
+
+]]
+
+---
+-- @usage
+-- nmap -p 80 --script http-jsonp-detection <target>
+--
+-- @output
+-- 80/tcp open  http    syn-ack
+-- | http-jsonp-detection: 
+-- |   JSONP endpoint found at /rest/contactsjp.php. Function name is callback
+-- |_  Callback function is completely controllable from the URL
+-- 
+--
+-- @xmloutput
+-- <table>
+-- <elem>JSONP endpoint found at /rest/contactsjp.php. Function name is callback</elem>
+-- <elem>Callback function is completely controllable from the URL</elem>
+-- </table>
+-- 
+-- @args http-jsonp-detection.path The URL path to request. The default path is "/".
+--- 
+
+author = {"Vinamra Bhatia"}
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {""} --to be figured out!
+
+portrule = shortport.port_or_service({80,443}, "http", "tcp")
+
+local callbacks = {"cb", "jsonp", "jsonpcallback", "jcb", "call"}
+
+--Checks the body and returns if valid json data is present in callback function
+local checkjson = function(body)
+   
+  local func, json_data
+  _, _, func, json_data = string.find(body, "(%S+)%((.*)%)")  --Check this!
+
+  --Check if the json_data is valid
+  --If valid, we have a JSONP endpoint with func as the function name
+
+  local status, json = json.parse(json_data)
+  return status, func
+end
+
+--Checks if the callback function is controllable from URL
+local callback_url = function(host, port, target)
+  local path, response, report
+  path = target .. "?callback=testing"
+  response = http.get(host, port, path)
+  if response and response.body and response.status==200 then
+
+    local status, func
+    status, func = checkjson(response.body)
+
+    if status == true then
+      if func == "testing" then
+        report = "Callback function is completely controllable from the URL"
+      else
+        local p = string.find(func1, "testing")
+        if p then 
+          report = "Callback function is partially controllable from URL"
+        end
+      end
+    end
+  end
+  return report
+end
+
+--The function tries to bruteforce through the most common callback variable
+local callback_bruteforce = function(host, port, target)
+  local response, path, report
+  for _,p in ipairs(callbacks) do 
+    path = target
+    path = path .. "?" .. p .. "=test"
+    response = http.get(host, port, path)
+    if response and response.body and response.status==200 then 
+
+      local status, func
+      status, func = checkjson(response.body)
+
+      if status == true then  
```

Comment:
Also remember that in this case you can also get endpoints that are partially controllable from URLs. This needs to be checked here as well.